libcaf_core/caf/message_builder.test.cpp

```diff
@@ -0,0 +1,84 @@
+// This file is part of CAF, the C++ Actor Framework. See the file LICENSE in
+// the main distribution directory for license terms and copyright or visit
+// https://github.com/actor-framework/actor-framework/blob/master/LICENSE.
+
+#include "caf/message_builder.hpp"
+
+#include "caf/test/caf_test_main.hpp"
+#include "caf/test/test.hpp"
+
+using namespace caf;
+
+TEST("message_builder can build messages incrementally with "
+     "default-constructor") {
+  message_builder builder;
+  check(builder.empty());
+  check(builder.to_message().empty());
+  check_eq(builder.size(), 0u);
+  SECTION("builder contains the values added by append") {
+    print_debug("after adding 1, the message is (1)");
+    {
+      builder.append(int32_t{1});
+      auto msg = builder.to_message();
+      check_eq(builder.size(), 1u);
+      check_eq(msg.types(), make_type_id_list<int32_t>());
+      check_eq(to_string(msg.types()), "[int32_t]");
+      check_eq(to_string(msg), "message(1)");
+    }
+    print_debug("after adding [2, 3], the message is (1, 2, 3)");
+    {
+      std::vector<int32_t> xs{2, 3};
+      builder.append(xs.begin(), xs.end());
+      check_eq(builder.size(), 3u);
+      auto msg = builder.to_message();
+      check_eq(msg.types(), (make_type_id_list<int32_t, int32_t, int32_t>()));
+      check_eq(to_string(msg.types()), "[int32_t, int32_t, int32_t]");
+      check_eq(to_string(msg), "message(1, 2, 3)");
+    }
+    SECTION("move_to_message produces the same message again") {
+      auto msg = builder.move_to_message();
+      check_eq(msg.types(), (make_type_id_list<int32_t, int32_t, int32_t>()));
+      check_eq(to_string(msg.types()), "[int32_t, int32_t, int32_t]");
+      check_eq(to_string(msg), "message(1, 2, 3)");
+    }
```

Comment:
I wouldn't nest this (as well as the other sections below) into the section that tests `append`.

Furthermore, the current setup ties the test for the builder to:
- the output format of `to_string`, which may change in the future
- the unrelated class `const_typed_message_view` (further below)

To test the individual values, please stick to interfaces provided by `message_builder` and `message`. The `move_to_message` test also needs to test that the builder actually no longer holds copies of old values.

```suggestion
    SECTION("move_to_message moves all elements into a message") {
      builder.append("foo"s);
      builder.append(make_counted<counted_int>(42));
      auto msg = builder.move_to_message();
      check_eq(msg.types(), (make_type_id_list<std::string, counted_int>()));
      if (check_eq(to_string(msg.types()), "[std::string, counted_int]")) {
        check_eq(msg.get_as<std::string>(0), "foo");
        const auto& ci = msg.get_as<counted_int_ptr>(1);
        check_eq(ci->value(), 42);
        check_eq(ci->get_reference_count(), 1u);
      }
    }
    SECTION("to_message copies all elements into a message") {
      builder.append("foo"s);
      builder.append(make_counted<counted_int>(42));
      check_eq(builder.size(), 2u);
      auto msg = builder.to_message();
      check_eq(msg.types(), make_type_id_list<std::string, counted_int>());
      if (check_eq(to_string(msg.types()), "[std::string, counted_int]")) {
        check_eq(msg.get_as<std::string>(0), "foo");
        const auto& ci = msg.get_as<counted_int_ptr>(1);
        check_eq(ci->value(), 42);
        check_eq(ci->get_reference_count(), 2u);
      }
    }
```

Whereas `counted_int` is something along those lines:

```cpp
class counted_int : public ref_counted {
public:
  explicit counted_int(int initial_value) : value_(initial_value) {
    // nop
  }

  int value() const {
    return value_;
  }

  void value(int new_value) {
    value_ = new_value;
  }

  template <class Inspector>
  friend bool inspect(Inspector& f, counted_int& x) {
    return f.object(x).fields(f.field("value", x.value_));
  }

private:
  int value_;
};

using counted_int_ptr = intrusive_ptr<counted_int>;

template <class Inspector>
bool inspect(Inspector& f, counted_int_ptr& x) {
  if constexpr (Inspector::is_loading) {
    if (!x)
      x = make_counted<counted_int>(0);
    return inspect(f, *x);
  } else {
    if (!x) {
      auto tmp = counted_int{0};
      return inspect(f, tmp);
    }
    return inspect(f, *x);
  }
}
```

(make sure to not forget the type ID block)

Reply:
In the updated implementation we will still be relying on the output format of to_string(). Do you think that we should look into updating that too?

```
     if (check_eq(to_string(msg.types()), "[std::string, counted_int]")) {
         check_eq(msg.get_as<std::string>(0), "foo");
         const auto& ci = msg.get_as<counted_int_ptr>(1);
         check_eq(ci->value(), 42);
         check_eq(ci->get_reference_count(), 2u);
      }
```